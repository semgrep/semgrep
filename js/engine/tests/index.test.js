const SEMGREP_PRO = process.env.SEMGREP_PRO;
const path = require("path");
const { EngineFactory } = require("../dist/index.cjs");

if (SEMGREP_PRO) {
  const java = require("../../languages/java/dist/index.cjs");
  const js = require("../../languages/typescript/dist/index.cjs");
}

const enginePromise = EngineFactory("./dist/semgrep-engine.wasm");

const engineExec = async () => {
  const engine = await enginePromise;
  if (SEMGREP_PRO) {
    const a = await java.ParserFactory();
    engine.addParser(a);
    const b = await js.ParserFactory();
    engine.addParser(b);
  }
  return engine;
};

describe("engine", () => {
  test("handles valid language", async () => {
    const engine = await enginePromise;
    expect(engine.lookupLang("java")).toEqual("java");
  });
  test("handles language aliases", async () => {
    const engine = await enginePromise;
    expect(engine.lookupLang("py")).toEqual("python");
  });
  test("handles invalid languages", async () => {
    const engine = await enginePromise;
    expect(engine.lookupLang("fake-language")).toBeNull();
  });
  test("detects an uninitialized language", async () => {
    const engine = await engineExec();
    engine.execute(
      "python",
      `${__dirname}/test-rule-python.json`,
      `${__dirname}`,
      [`dirname}/../../languages/python/tests/test.py`]
    );
    expect(engine.isMissingLanguages()).toBe(true);
    expect(engine.getMissingLanguages()).toEqual(["python"]);
    engine.clearMissingLanguages();
    expect(engine.isMissingLanguages()).toBe(false);
  });
});

describe("yaml parser", () => {
  test("parses a simple pattern", async () => {
    const engine = await engineExec();
    const rulePath = path.resolve(`${__dirname}/test-rule-yaml.json`);
    const targetPath = path.resolve(`${__dirname}/test.yaml`);
    const result = JSON.parse(
      engine
        .execute("yaml", rulePath, `${__dirname}`, [targetPath])
        .replaceAll(rulePath, "test-rule-yaml.json")
        .replaceAll(targetPath, "test.yaml")
        .replaceAll("PRO", "OSS")
    );
    expect(result).toMatchSnapshot();
  });
  test("parses a pattern with pattern-regex", async () => {
    const engine = await enginePromise;
    const rulePath = path.resolve(`${__dirname}/test-rule-yaml-regex.json`);
    const targetPath = path.resolve(`${__dirname}/test.yaml`);

    const result = JSON.parse(
      engine
        .execute("yaml", rulePath, `${__dirname}`, [targetPath])
        .replaceAll(rulePath, "test-rule-yaml.json")
        .replaceAll(targetPath, "test.yaml")
        .replaceAll("PRO", "OSS")
    );

    expect(result).toMatchSnapshot();
  });
});

if (SEMGREP_PRO) {
  describe("deep", () => {
    test("parses a pattern", async () => {
      const engine = await engineExec();
      let paths = [
        path.resolve(`${__dirname}/A.java`),
        path.resolve(`${__dirname}/B.java`),
        path.resolve(`${__dirname}/C.java`),
      ];
      const result = JSON.parse(
        engine
          .execute(
            "java",
            `${__dirname}/test-rule-java.json`,
            `${__dirname}`,
            paths
          )
          .replaceAll(path[0], "A.java")
          .replaceAll(path[1], "B.java")
          .replaceAll(path[2], "C.java")
      );
      expect(result).toMatchSnapshot();
    });
  });
}
