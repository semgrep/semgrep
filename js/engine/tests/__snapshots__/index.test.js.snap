// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`yaml parser parses a pattern with pattern-regex 1`] = `
{
  "errors": [],
  "explanations": [
    {
      "children": [],
      "loc": {
        "end": {
          "col": 30,
          "line": 6,
          "offset": 99,
        },
        "path": "test-rule-yaml-regex.json",
        "start": {
          "col": 24,
          "line": 6,
          "offset": 93,
        },
      },
      "matches": [
        {
          "check_id": "test",
          "end": {
            "col": 4,
            "line": 1,
            "offset": 3,
          },
          "extra": {
            "engine_kind": "OSS",
            "message": "test",
            "metavars": {},
            "validation_state": "NO_VALIDATOR",
          },
          "path": "test.yaml",
          "start": {
            "col": 1,
            "line": 1,
            "offset": 0,
          },
        },
      ],
      "op": [
        "XPat",
        "^fo+",
      ],
    },
  ],
  "paths": {
    "scanned": [
      "test.yaml",
    ],
  },
  "results": [
    {
      "check_id": "test",
      "end": {
        "col": 4,
        "line": 1,
        "offset": 3,
      },
      "extra": {
        "engine_kind": "OSS",
        "fingerprint": "<MASKED>",
        "is_ignored": false,
        "lines": "foo: bar",
        "message": "test",
        "metadata": {},
        "metavars": {},
        "severity": "ERROR",
        "validation_state": "NO_VALIDATOR",
      },
      "path": "test.yaml",
      "start": {
        "col": 1,
        "line": 1,
        "offset": 0,
      },
    },
  ],
  "skipped_rules": [],
  "version": "<MASKED>",
}
`;

exports[`yaml parser parses a simple pattern 1`] = `
{
  "errors": [],
  "explanations": [
    {
      "children": [],
      "loc": {
        "end": {
          "col": 27,
          "line": 6,
          "offset": 96,
        },
        "path": "test-rule-yaml.json",
        "start": {
          "col": 18,
          "line": 6,
          "offset": 87,
        },
      },
      "matches": [
        {
          "check_id": "test",
          "end": {
            "col": 9,
            "line": 1,
            "offset": 8,
          },
          "extra": {
            "engine_kind": "OSS",
            "message": "test",
            "metavars": {
              "$X": {
                "abstract_content": "bar",
                "end": {
                  "col": 9,
                  "line": 1,
                  "offset": 8,
                },
                "start": {
                  "col": 6,
                  "line": 1,
                  "offset": 5,
                },
              },
            },
            "validation_state": "NO_VALIDATOR",
          },
          "path": "test.yaml",
          "start": {
            "col": 1,
            "line": 1,
            "offset": 0,
          },
        },
      ],
      "op": [
        "XPat",
        "foo: $X",
      ],
    },
  ],
  "paths": {
    "scanned": [
      "test.yaml",
    ],
  },
  "results": [
    {
      "check_id": "test",
      "end": {
        "col": 9,
        "line": 1,
        "offset": 8,
      },
      "extra": {
        "engine_kind": "OSS",
        "fingerprint": "<MASKED>",
        "is_ignored": false,
        "lines": "foo: bar",
        "message": "test",
        "metadata": {},
        "metavars": {
          "$X": {
            "abstract_content": "bar",
            "end": {
              "col": 9,
              "line": 1,
              "offset": 8,
            },
            "start": {
              "col": 6,
              "line": 1,
              "offset": 5,
            },
          },
        },
        "severity": "ERROR",
        "validation_state": "NO_VALIDATOR",
      },
      "path": "test.yaml",
      "start": {
        "col": 1,
        "line": 1,
        "offset": 0,
      },
    },
  ],
  "skipped_rules": [],
  "version": "<MASKED>",
}
`;
